src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceHelpers.cs

```diff
@@ -96,28 +103,243 @@ public static async Task<Location> GetLocationInGeneratedSourceAsync(SymbolKey s
+                        return null;
+
+                    // If we're looking for a type, we're already here
+                    if (symbol.Kind == SymbolKind.NamedType)
+                        return typeDefHandle;
+
+                    // Now we have the type, lets find the actual symbol
+                    return FindEntityHandle(md, typeDefinition.Value, symbol);
+                }
+            }
+            catch (Exception ex) when (IOUtilities.IsNormalIOException(ex))
+            {
+            }
+
+            return null;
+        }
+
+        private static string? GetExportedTypeAssemblyName(MetadataReader md, INamedTypeSymbol typeSymbol)
+        {
+            foreach (var eth in md.ExportedTypes)
+            {
+                var et = md.GetExportedType(eth);
+                if (!et.IsForwarder)
+                    continue;
+
+                if (md.StringComparer.Equals(et.Name, typeSymbol.MetadataName) &&
+                    md.StringComparer.Equals(et.Namespace, typeSymbol.ContainingNamespace.MetadataName))
+                {
+                    var handle = et.Implementation;
+                    // For nested types, the exported type references another exported type, so keep looking up
+                    // until we get something else
+                    while (handle.Kind == HandleKind.ExportedType)
+                    {
+                        handle = md.GetExportedType((ExportedTypeHandle)handle).Implementation;
+                    }
+
+                    Debug.Assert(handle.Kind == HandleKind.AssemblyReference);
+
+                    var ar = md.GetAssemblyReference((AssemblyReferenceHandle)handle);
+                    return md.GetString(ar.Name);
+                }
+            }
+
+            return null;
+        }
+
+        private static TypeDefinition? GetTypeDefinition(MetadataReader md, INamedTypeSymbol typeSymbol, out TypeDefinitionHandle typeDefinitionHandle)
+        {
+            foreach (var tdh in md.TypeDefinitions)
+            {
+                var td = md.GetTypeDefinition(tdh);
+                if (md.StringComparer.Equals(td.Name, typeSymbol.MetadataName) &&
+                    md.StringComparer.Equals(td.Namespace, typeSymbol.ContainingNamespace.MetadataName))
+                {
+                    typeDefinitionHandle = tdh;
+                    return td;
+                }
+            }
+
+            typeDefinitionHandle = default;
+            return null;
+        }
+
+        private static EntityHandle? FindEntityHandle(MetadataReader md, TypeDefinition typeDefinition, ISymbol symbol)
+        {
+            switch (symbol.Kind)
+            {
+                case SymbolKind.Method:
+                    foreach (var mdh in typeDefinition.GetMethods())
+                    {
+                        var methodDef = md.GetMethodDefinition(mdh);
+                        if (md.StringComparer.Equals(methodDef.Name, symbol.MetadataName))
```

Comment:
This presumably needs more work, to check parameter count, type parameter count etc. Checking parameter types match is presumably extremely annoying...

Reply:
We already have this logic implemented SymbolKey. Can you create an empty compilation with the impl assembly as a reference and ask SymbolKey for `symbol` to resolve in it?